pkg/controller/master/node/node_down_controller.go

```diff
@@ -3,278 +3,379 @@ package node
 		}
-		names = append(names, pvc.Spec.VolumeName)
 	}
-	return names, nil
+	if _, err := h.nodes.UpdateStatus(nodeCpy); err != nil {
+		return err
+	}
+	return nil
 }
 
-// deleteVolumeAttachments deletes all volume attachments of PVCs before post-host failure,
-// due to the long controller-manager CSI processing period,
-// causing the VM to restart to wait a long time.
-// Then we delete VolumeAttachments directly.
-// ref: https://github.com/harvester/harvester/issues/4049
-func (h *nodeDownHandler) deleteVolumeAttachments(nodeName string, pvcSet set.Set[string]) error {
-	if pvcSet.Len() == 0 {
-		return nil
+func (h *nodeDownHandler) fetchVMForceResetPolicy() (*settings.VMForceResetPolicy, error) {
+	vmForceResetPolicy, err := settings.DecodeVMForceResetPolicy(settings.VMForceResetPolicySet.Get())
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode VMForceResetPolicy setting: %w", err)
 	}
+	return vmForceResetPolicy, nil
+}
 
-	volumeAttachments, err := h.vaCache.List(labels.Everything())
-	if err != nil {
-		return err
+func (h *nodeDownHandler) tryWaitForExtraTimeout(node *corev1.Node, extraTime int64) (bool, error) {
+	// get VMForceResetPolicy setting
+	targetInterval := time.Duration(extraTime) * time.Second
+
+	// check if we wait enough (targetInterval) after the kubelet becomes not ready
+	cond := getNodeCondition(node.Status.Conditions, corev1.NodeReady)
+	if cond != nil && time.Since(cond.LastTransitionTime.Time) < targetInterval {
+		deadline := cond.LastTransitionTime.Add(targetInterval)
+		logrus.Debugf("Enqueue node event again at %v", deadline)
+		h.nodes.EnqueueAfter(node.Name, time.Until(deadline))
+		return false, nil
 	}
-	for i := range volumeAttachments {
-		va := volumeAttachments[i]
 
-		if va.DeletionTimestamp != nil {
-			continue
-		}
-		if va.Spec.NodeName != nodeName {
-			continue
-		}
-		if va.Spec.Attacher != longhorntypes.LonghornDriverName {
-			continue
-		}
-		if va.Spec.Source.PersistentVolumeName == nil {
-			continue
-		}
-		if !pvcSet.Has(*va.Spec.Source.PersistentVolumeName) {
-			continue
-		}
+	return true, nil
+}
 
-		if err := h.vas.Delete(va.Name, &metav1.DeleteOptions{}); err != nil {
-			if errors.IsNotFound(err) {
-				continue
-			}
-			return err
+func (h *nodeDownHandler) tryWaitForVMMigration(node *corev1.Node, vmMigrationTimeout int64) (bool, error) {
+	// check if we wait enough (vmMigrationTimeout) after the kubevirtDrainTaint is added
+	cond := getNodeCondition(node.Status.Conditions, HarvesterNodeCondDrained)
+	if cond != nil && cond.Status == corev1.ConditionTrue {
+		// already drained
+		return true, nil
+	}
+	targetInterval := time.Duration(vmMigrationTimeout) * time.Second
+	if cond == nil {
+		// means we need to change the condition to unreachable (add a new one or update the existing one)
+		if err := h.setHarvesterNodeDrainedCondition(node, corev1.ConditionFalse, HarvesterNodeFailureCondReasonUnreachable, HarvesterNodeFailureCondMsgUnreachable); err != nil {
```

Comment:
Re: the `reason` for this condition (here and in line 131), normally, we specify the reason (not the cause) for the true/false status of a condition. In our case, the node drained condition is false because draining is still in progress or the timeout hasn't expired. Conversely in line 131, the node drained condition is true because draining is completed or the timeout has expired. 

Node failure is the cause that triggers this condition (which can be the `message`), not the reason why the condition is in a certain state.